### PR TITLE
fix: Ensure that each request can only fail once.

### PR DESCRIPTION
**Pull request by kinyoklion**
On WSL with no network, or running docker without a network, the 'error' and 'timeout' callbacks would both be called. The timeout should only happen after the specified timeout, 5 minutes in the node SDK, but happens immediately under either of those conditions.

This change ensures that each request can only fail one time.

### CODE REVIEW

**Comment by kinyoklion** [comment]
I think there may be another bug where you call close while a reconnect is already scheduled, but I will handle that in another PR.

File: lib/eventsource.js [hunk 1]
```diff
@@ -324,12 +348,14 @@ function EventSource (url, eventSourceInitDict) {
...
-      failed({ message: err.message })
+      failOnce({ message: err.message })
     })
 
     req.on('timeout', function () {
-      failed({ message: 'Read timeout, received no data in ' + config.readTimeoutMillis +
+      failOnce({ message: 'Read timeout, received no data in ' + config.readTimeoutMillis +
           'ms, assuming connection is dead' })
+      // Timeout doesn't mean that the request is cancelled, just that it has elapsed the timeout.
+      destroyRequest()
```
Comment (kinyoklion) [inline lib/eventsource.js:358]: Being more careful about destroying requests will help to ensure applications are not held open by the event source.

File: example/eventsource-polyfill.js [hunk 2]
```diff
@@ -67,27 +67,27 @@
...
-		g = window;
-}
-
-// g can still be undefined, but nothing to do about it...
-// We return undefined, instead of nothing here, so it's
-// easier to handle this case. if(!global) { ...}
-
-module.exports = g;
+var g;
+
```
Comment (cwaldren-ld) [inline example/eventsource-polyfill.js:71]: this just formatting or was there a change?

File: example/eventsource-polyfill.js [hunk 3]
```diff
@@ -7264,6 +7264,19 @@ function hasBom (buf) {
   })
 }
 
+/**
+ * Wrap a callback to ensure it can only be called once.
+ */
+function once(cb) {
+  let called = false
```
Comment (cwaldren-ld) [inline example/eventsource-polyfill.js:7271]: It sure is nice to be single threaded

File: example/eventsource-polyfill.js [hunk 4]
```diff
@@ -12056,4 +12080,4 @@ module.exports = function isBuffer(arg) {
 }
 
 /***/ })
-/******/ ]);
+/******/ ]);
```
Comment (cwaldren-ld) [inline example/eventsource-polyfill.js:12083]: newline

**Review by cwaldren-ld — APPROVED** [review]

File: example/eventsource-polyfill.js (hunk 2 above)
Comment (kinyoklion) [reply example/eventsource-polyfill.js:71]: This file is auto-generated based on the other files. I am not sure the cause of the diff here. It looks like most of the formatting is even the same.

File: example/eventsource-polyfill.js (hunk 4 above)
Comment (kinyoklion) [reply example/eventsource-polyfill.js:12083]: I'm not sure if it was manually edited before, but it generated this way.

File: example/eventsource-polyfill.js (hunk 3 above)
Comment (kinyoklion) [reply example/eventsource-polyfill.js:7271]: It is very nice indeed. (at least the illusion of a single thread.)

File: lib/eventsource.js [hunk 5]
```diff
@@ -27,6 +27,19 @@ function hasBom (buf) {
...
+/**
+ * Wrap a callback to ensure it can only be called once.
+ */
+function once(cb) {
+  let called = false
+  return (...params) => {
+    if(!called) {
+      called = true
+      cb(...params)
+    }
```
Comment (yusinto) [inline lib/eventsource.js:39]: Is there a logger util in this library? It's helpful to see logs when something is ignored.

Comment (kinyoklion) [reply lib/eventsource.js:39]: There is not.

**Pull request merged**